Lucene error - IndexOutOfBoundsException accessing norms

Came across this bug today on master and I have been unable to reduce to a simple case.
Lucene throws an exception accessing norms only when given the following conditions:
* Only when query is for term A OR term B (doesn't fail with either of the terms independently)
* Only when using top docs (works OK with hits)
* Only when the docs matching the query sit in an index with other docs.

The gist with the data and queries to reproduce this is here: 
https://gist.github.com/markharwood/3bdf60db7887ec5eb7a0a8fd7074cfae
I have reduced the number of docs ingested to the smallest set of fields and the smallest number of records I could before the error is revealed.

Error thrown is 

	Caused by: java.lang.IndexOutOfBoundsException: 2147483647
		at java.nio.DirectByteBuffer.get(DirectByteBuffer.java:253) ~[?:1.8.0_144]
		at org.apache.lucene.store.ByteBufferGuard.getByte(ByteBufferGuard.java:118) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.store.ByteBufferIndexInput$SingleBufferImpl.readByte(ByteBufferIndexInput.java:385) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.codecs.lucene70.Lucene70NormsProducer$2.longValue(Lucene70NormsProducer.java:218) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.similarities.BM25Similarity$BM25DocScorer.score(BM25Similarity.java:253) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.TermScorer.score(TermScorer.java:66) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.TopScoreDocCollector$SimpleTopScoreDocCollector$1.collect(TopScoreDocCollector.java:64) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.elasticsearch.search.aggregations.metrics.tophits.TopHitsAggregator$1.collect(TopHitsAggregator.java:132) ~[1/:?]
		at org.elasticsearch.search.aggregations.LeafBucketCollector.collect(LeafBucketCollector.java:82) ~[1/:?]
		at org.apache.lucene.search.MultiCollector$MultiLeafCollector.collect(MultiCollector.java:174) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BooleanScorer.scoreDocument(BooleanScorer.java:189) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BooleanScorer.scoreMatches(BooleanScorer.java:202) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BooleanScorer.scoreWindowIntoBitSetAndReplay(BooleanScorer.java:216) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BooleanScorer.scoreWindowMultipleScorers(BooleanScorer.java:260) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BooleanScorer.scoreWindow(BooleanScorer.java:305) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BooleanScorer.score(BooleanScorer.java:317) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:39) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:658) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:186) ~[1/:?]
		at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:462) ~[lucene-core-7.1.0-snapshot-f33ed4ba12a.jar:7.1.0-snapshot-f33ed4ba12a f33ed4ba12aaf215628d010daaa0e271b8ab3d1f - mvg - 2017-10-02 17:18:30]
		at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:272) ~[1/:?]
		at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:110) ~[1/:?]
		at org.elasticsearch.indices.IndicesService.lambda$16(IndicesService.java:1122) ~[1/:?]
		at org.elasticsearch.indices.IndicesService.lambda$17(IndicesService.java:1175) ~[1/:?]
		at org.elasticsearch.indices.IndicesRequestCache$Loader.load(IndicesRequestCache.java:160) ~[1/:?]
		at org.elasticsearch.indices.IndicesRequestCache$Loader.load(IndicesRequestCache.java:1) ~[1/:?]
		at org.elasticsearch.common.cache.Cache.computeIfAbsent(Cache.java:412) ~[1/:?]
		at org.elasticsearch.indices.IndicesRequestCache.getOrCompute(IndicesRequestCache.java:116) ~[1/:?]
		at org.elasticsearch.indices.IndicesService.cacheShardLevelResult(IndicesService.java:1181) ~[1/:?]
		at org.elasticsearch.indices.IndicesService.loadIntoContext(IndicesService.java:1121) ~[1/:?]
		at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:300) ~[1/:?]
		at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:335) ~[1/:?]
		at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:311) ~[1/:?]

--- a/core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregator.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregator.java
@@ -20,6 +20,8 @@
 package org.elasticsearch.search.aggregations.metrics.tophits;
 
 import com.carrotsearch.hppc.LongObjectHashMap;
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.search.FieldDoc;
 import org.apache.lucene.search.LeafCollector;
@@ -93,6 +95,9 @@ public class TopHitsAggregator extends MetricsAggregator {
             public void setScorer(Scorer scorer) throws IOException {
                 this.scorer = scorer;
                 super.setScorer(scorer);
+                for (ObjectCursor<LeafCollector> cursor : leafCollectors.values()) {
+                    cursor.value.setScorer(scorer);
+                }
             }
 
             @Override

--- a/core/src/test/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorTests.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/metrics/tophits/TopHitsAggregatorTests.java
@@ -21,15 +21,22 @@ package org.elasticsearch.search.aggregations.metrics.tophits;
 import org.apache.lucene.analysis.core.KeywordAnalyzer;
 import org.apache.lucene.document.Document;
 import org.apache.lucene.document.Field;
+import org.apache.lucene.document.Field.Store;
 import org.apache.lucene.document.SortedSetDocValuesField;
+import org.apache.lucene.document.StringField;
 import org.apache.lucene.index.DirectoryReader;
 import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.IndexWriter;
 import org.apache.lucene.index.RandomIndexWriter;
+import org.apache.lucene.index.Term;
 import org.apache.lucene.queryparser.classic.QueryParser;
+import org.apache.lucene.search.BooleanClause.Occur;
+import org.apache.lucene.search.BooleanQuery;
 import org.apache.lucene.search.IndexSearcher;
 import org.apache.lucene.search.MatchAllDocsQuery;
 import org.apache.lucene.search.MatchNoDocsQuery;
 import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.util.BytesRef;
 import org.elasticsearch.index.mapper.KeywordFieldMapper;
@@ -39,6 +46,7 @@ import org.elasticsearch.index.mapper.UidFieldMapper;
 import org.elasticsearch.search.SearchHits;
 import org.elasticsearch.search.aggregations.Aggregation;
 import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilders;
 import org.elasticsearch.search.aggregations.AggregatorTestCase;
 import org.elasticsearch.search.aggregations.bucket.terms.Terms;
 import org.elasticsearch.search.sort.SortOrder;
@@ -147,5 +155,48 @@ public class TopHitsAggregatorTests extends AggregatorTestCase {
             document.add(new SortedSetDocValuesField("string", new BytesRef(stringValue)));
         }
         return document;
+    }
+
+    public void testSetScorer() throws Exception {
+        Directory directory = newDirectory();
+        IndexWriter w = new IndexWriter(directory, newIndexWriterConfig()
+                // only merge adjacent segments
+                .setMergePolicy(newLogMergePolicy()));
+        // first window (see BooleanScorer) has matches on one clause only
+        for (int i = 0; i < 2048; ++i) {
+            Document doc = new Document();
+            doc.add(new StringField("_id", Uid.encodeId(Integer.toString(i)), Store.YES));
+            if (i == 1000) { // any doc in 0..2048
+                doc.add(new StringField("string", "bar", Store.NO));
+            }
+            w.addDocument(doc);
+        }
+        // second window has matches in two clauses
+        for (int i = 0; i < 2048; ++i) {
+            Document doc = new Document();
+            doc.add(new StringField("_id", Uid.encodeId(Integer.toString(2048 + i)), Store.YES));
+            if (i == 500) { // any doc in 0..2048
+                doc.add(new StringField("string", "baz", Store.NO));
+            } else if (i == 1500) {
+                doc.add(new StringField("string", "bar", Store.NO));
+            }
+            w.addDocument(doc);
+        }
+
+        w.forceMerge(1); // we need all docs to be in the same segment
+
+        IndexReader reader = DirectoryReader.open(w);
+        w.close();
+
+        IndexSearcher searcher = new IndexSearcher(reader);
+        Query query = new BooleanQuery.Builder()
+                .add(new TermQuery(new Term("string", "bar")), Occur.SHOULD)
+                .add(new TermQuery(new Term("string", "baz")), Occur.SHOULD)
+                .build();
+        AggregationBuilder agg = AggregationBuilders.topHits("top_hits");
+        TopHits result = searchAndReduce(searcher, query, agg, STRING_FIELD_TYPE);
+        assertEquals(3, result.getHits().totalHits);
+        reader.close();
+        directory.close();
     }
 }

--- a/test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java
@@ -91,6 +91,7 @@ import static org.mockito.Mockito.when;
 public abstract class AggregatorTestCase extends ESTestCase {
     private static final String NESTEDFIELD_PREFIX = "nested_";
     private List<Releasable> releasables = new ArrayList<>();
+    private static final String TYPE_NAME = "type";
 
     /** Create a factory for the given aggregation builder. */
     protected AggregatorFactory<?> createAggregatorFactory(AggregationBuilder aggregationBuilder,
@@ -104,6 +105,7 @@ public abstract class AggregatorTestCase extends ESTestCase {
         MapperService mapperService = mapperServiceMock();
         when(mapperService.getIndexSettings()).thenReturn(indexSettings);
         when(mapperService.hasNested()).thenReturn(false);
+        when(mapperService.types()).thenReturn(Collections.singleton(TYPE_NAME));
         when(searchContext.mapperService()).thenReturn(mapperService);
         IndexFieldDataService ifds = new IndexFieldDataService(indexSettings,
                 new IndicesFieldDataCache(Settings.EMPTY, new IndexFieldDataCache.Listener() {
@@ -115,7 +117,7 @@ public abstract class AggregatorTestCase extends ESTestCase {
             }
         });
 
-        SearchLookup searchLookup = new SearchLookup(mapperService, ifds::getForField, new String[]{"type"});
+        SearchLookup searchLookup = new SearchLookup(mapperService, ifds::getForField, new String[]{TYPE_NAME});
         when(searchContext.lookup()).thenReturn(searchLookup);
 
         QueryShardContext queryShardContext = queryShardContextMock(mapperService, fieldTypes, circuitBreakerService);

